src/lib/ledger_catchup/ledger_catchup.ml

```diff
@@ -120,118 +129,185 @@ module Make (Inputs : Inputs.S) :
+                    sprintf
+                      !"Peer %{sexp:Network_peer.Peer.t} sent us bad proof"
+                      peer
+                  in
+                  ignore
+                    Trust_system.(
+                      record trust_system logger peer.host
+                        Actions.(Violated_protocol, Some (error_msg, []))) ;
+                  Or_error.error_string error_msg
+            in
+            List.fold_until hashes ~init:[]
+              ~f:(fun acc hash ->
+                if Transition_frontier.find frontier hash |> Option.is_some
+                then Continue_or_stop.Stop (Ok (peer, acc))
+                else Continue_or_stop.Continue (hash :: acc) )
+              ~finish:(fun _ ->
+                Or_error.errorf
+                  !"Peer %{sexp:Network_peer.Peer.t} moves too fast"
+                  peer ) )
+
+  let verify_against_hashes transitions hashes =
+    List.length transitions = List.length hashes
+    && List.for_all2_exn transitions hashes ~f:(fun transition hash ->
+           State_hash.equal
+             ( External_transition.protocol_state transition
+             |> Protocol_state.hash )
+             hash )
+
+  let rec partition size = function
+    | [] ->
+        []
+    | ls ->
+        let sub, rest = List.split_n ls size in
+        sub :: partition size rest
 
-  let take_while_map_result_rev ~f list =
+  (* returns a list of transitions with old ones comes first *)
+  let download_transitions ~logger ~trust_system ~network ~num_peers
+      ~preferred_peer ~maximum_download_size ~hashes_of_missing_transitions =
+    let random_peers = Network.random_peers network num_peers in
+    Deferred.Or_error.List.concat_map
+      (partition maximum_download_size hashes_of_missing_transitions)
+      ~how:`Sequential ~f:(fun hashes ->
+        Deferred.Or_error.find_map_ok (preferred_peer :: random_peers)
+          ~f:(fun peer ->
+            match%bind Network.get_transition_chain network peer hashes with
+            | None ->
+                Deferred.Or_error.errorf
+                  !"Peer %{sexp:Network_peer.Peer.t} did not have the \
+                    transitions we requested"
+                  peer
+            | Some transitions ->
+                if not @@ verify_against_hashes transitions hashes then (
+                  let error_msg =
+                    sprintf
+                      !"Peer %{sexp:Network_peer.Peer.t} returned a list that \
+                        is different from the one that is requested."
+                      peer
+                  in
+                  Trust_system.(
+                    record trust_system logger peer.host
+                      Actions.(Violated_protocol, Some (error_msg, [])))
+                  |> don't_wait_for ;
+                  Deferred.Or_error.error_string error_msg )
+                else
+                  Deferred.Or_error.return
+                  @@ List.map transitions ~f:(fun transition ->
```

Comment:
Seems like you are computing the hash of an external_transition twice, once in the statement for creating `transition_with_hash` and in the if statement `verify_again_hashes`. Let's compute the hash once. Hashing is slow.

Reply:
Sure, good catch.